### PR TITLE
PixelPaint: Add Text tool

### PR DESCRIPTION
**Pull request by tslater2006**
This adds a Text Tool to PixelPaint. Fonts are selected by the GUI::FontPicker dialog. User's can click to start a new text entry, the color of the text will match the selected primary/secondary colors depending on if a left or right click was used. Font can be changed during a text entry and take effect immediately. Text Tool allows a text entry to be dragged around while it is being edited. Once a text entry is commited to the layer (via the Return key) it cannot be edited anymore.

Currently there is an issue with single letter action shortcuts, if you type one of those letters (like 'c' for example) it will trigger the action and cause you to leave the Text tool. There are a couple of PRs around solving this (PR #15788 and PR #15773). Once a solution is settled on/merged I will rebase and make any adjustments necessary to the Text Tool. Once resolved I will flip this to a full PR instead of draft.

### CODE REVIEW

File: Userland/Applications/PixelPaint/Tools/TextTool.cpp [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+{
+    if (!m_is_typing)
+        return;
+
+    GUI::Painter painter(*m_editor);
+    painter.add_clip_rect(event.rect());
+    painter.translate(editor_layer_location(*layer));
+    auto typed_text = m_text_editor->text();
+    auto text_width = max<int>(m_selected_font->width(typed_text), m_selected_font->width(" "sv));
+    auto text_height = m_selected_font->preferred_line_height() * max<int>((int)m_text_editor->line_count(), 1);
```
Comment (GMTA) [inline Userland/Applications/PixelPaint/Tools/TextTool.cpp:130]: Please use C++-style casts (`static_cast<int>`) instead of `(int)` everywhere

Comment (tslater2006) [reply Userland/Applications/PixelPaint/Tools/TextTool.cpp]: Resolved in latest push :)

**Comment by tslater2006** [comment]
Rebased on latest master to fix CI build failure. Build failure was due to Lasso Tool needing to be updated to remove the const modified on the on_keydown() like other tools.

File: Userland/Applications/PixelPaint/Tools/TextTool.cpp [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+#include <LibGUI/Action.h>
+#include <LibGUI/BoxLayout.h>
+#include <LibGUI/Button.h>
+#include <LibGUI/Label.h>
+#include <LibGUI/Menu.h>
+#include <LibGUI/Painter.h>
+#include <LibGUI/TextBox.h>
+#include <LibGfx/Bitmap.h>
+#include <LibGfx/Palette.h>
+namespace PixelPaint {
```
Comment (AtkinsSJ) [inline Userland/Applications/PixelPaint/Tools/TextTool.cpp:20]: Nit-pick: A blank line between the includes and the namespace please.

File: Userland/Applications/PixelPaint/Tools/TextTool.cpp [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+        // Draw selected text for each line...
+        auto selection_start_line = selection.normalized().start().line();
+        auto selection_end_line = selection.normalized().end().line();
+
+        for (auto i = selection_start_line; i <= selection_end_line; ++i) {
+
+            auto start_col = 0;
+            auto end_col = 0;
+
+            // first line of selection
```
Comment (awesomekling) [inline Userland/Applications/PixelPaint/Tools/TextTool.cpp:160]: Please write comments as proper sentences, i.e "First line of selection."

File: Userland/Applications/PixelPaint/Tools/TextTool.cpp [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+
+    // blinking cursor
+    if (m_cursor_state) {
+        auto cursor_text_prefix = m_text_editor->line(m_text_editor->cursor().line()).to_utf8().substring(0, m_text_editor->cursor().column());
+        auto cursor_x_offset = static_cast<int>((m_selected_font->width(cursor_text_prefix) + 1) * m_editor->scale());
+        auto cursor_y_offset = static_cast<int>((m_text_editor->cursor().line() * (m_selected_font->pixel_size() + 4)) * m_editor->scale());
+
+        auto cursor_width = max<int>(1 * m_editor->scale(), 1);
+        auto cursor_height = (m_selected_font->pixel_size() + 4) * m_editor->scale();
+        auto cursor_rect = Gfx::Rect<int>(text_location.x() + cursor_x_offset, text_location.y() + cursor_y_offset, cursor_width, cursor_height);
```
Comment (awesomekling) [inline Userland/Applications/PixelPaint/Tools/TextTool.cpp:220]: This is a heck of a lot of work to get a cursor rect. Is there any way we can leverage the work already done by `m_text_editor` to simplify this?

**Review by awesomekling — CHANGES_REQUESTED** [review]

**Comment by AtkinsSJ** [comment]
Noticed while trying this out that the text draws over the rulers:
image

**Review by AtkinsSJ — CHANGES_REQUESTED** [review]

File: Userland/Applications/PixelPaint/Tools/Tool.h [hunk 5]
```diff
@@ -64,6 +64,7 @@ class Tool {
     virtual bool on_keydown(GUI::KeyEvent const&);
     virtual void on_keyup(GUI::KeyEvent&) { }
     virtual void on_tool_activation() { }
+    virtual void on_tool_exit() { }
```
Comment (AtkinsSJ) [inline Userland/Applications/PixelPaint/Tools/Tool.h:67]: `on_tool_deactivation` would be more clearly related to `on_tool_activation`

File: Userland/Applications/PixelPaint/Tools/TextTool.h [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+#include "Tool.h"
+#include <LibCore/Timer.h>
+#include <LibGUI/ActionGroup.h>
+#include <LibGUI/Label.h>
+#include <LibGUI/Painter.h>
+#include <LibGUI/TextEditor.h>
+#include <LibGUI/TextPosition.h>
+
+namespace PixelPaint {
+class TextTool;
```
[the remark this answers is not in the document]
Comment (tslater2006) [reply Userland/Applications/PixelPaint/Tools/TextTool.h:18]: This is used in TextToolEditor's handle_keyevent, for the `Badge<TextTool>` parameter. TextToolEditor is declared above TextTool so I put the forward declare in.

File: Userland/Applications/PixelPaint/Tools/Tool.h (hunk 5 above)
Comment (tslater2006) [reply Userland/Applications/PixelPaint/Tools/Tool.h]: Thanks for the suggestion. I've changed this in latest push.

File: Userland/Applications/PixelPaint/Tools/TextTool.cpp [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+    m_timer->set_interval(500);
+}
+
+void TextTool::on_tool_exit()
+{
+    reset_tool();
+}
+void TextTool::on_mousemove(Layer*, MouseEvent& event)
+{
+
```
[the remark this answers is not in the document]
Comment (tslater2006) [reply Userland/Applications/PixelPaint/Tools/TextTool.cpp]: Fixed

File: Userland/Applications/PixelPaint/Tools/TextTool.cpp (hunk 2 above)
Comment (tslater2006) [reply Userland/Applications/PixelPaint/Tools/TextTool.cpp:21]: Fixed

File: Userland/Applications/PixelPaint/Tools/TextTool.h [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    virtual GUI::Widget* get_properties_widget() override;
+
+private:
+    virtual StringView tool_name() const override { return "Text Tool"sv; }
+    void apply_text_to_layer();
+    void reset_tool();
+
+    RefPtr<GUI::Widget> m_properties_widget;
+    RefPtr<GUI::Label> m_font_label;
+    RefPtr<Core::Timer> m_timer;
```
[the remark this answers is not in the document]
Comment (tslater2006) [reply Userland/Applications/PixelPaint/Tools/TextTool.h]: Yep, much better. Still getting used to the whole "name things exactly what they are" idea (which I do love by the way).

File: Userland/Applications/PixelPaint/Tools/TextTool.h [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    RefPtr<PixelPaint::TextToolEditor> m_text_editor;
+    Gfx::IntPoint m_add_text_position { 0, 0 };
+    RefPtr<Gfx::Font> m_selected_font;
+    bool m_is_typing { false };
```
[the remark this answers is not in the document]
Comment (tslater2006) [reply Userland/Applications/PixelPaint/Tools/TextTool.h]: For now I've gone with "m_text_input_is_active", but I do worry that "text input" might be read wrong (as in like an object, not a state). My other thought was "m_accepting_text_input" or something, but I think this is good enough for now and at least better than it was.

File: Userland/Applications/PixelPaint/Tools/TextTool.cpp [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+            auto selection_y_offset = static_cast<int>((m_selected_font->pixel_size() + 4) * i);
+
+            auto selection_rect = Gfx::IntRect(selection_x_offset, selection_y_offset, selected_width, m_selected_font->preferred_line_height());
+            text_painter.fill_rect(selection_rect, m_text_editor->palette().selection());
+            text_painter.draw_text(selection_rect, selected_string, Gfx::TextAlignment::TopLeft, m_text_editor->palette().selection_text());
+        }
+    }
+
+    /* auto scaled_bitmap = text_bitmap->scaled(m_editor->scale(), m_editor->scale()).release_value();
+    painter.blit({ text_location.x(), text_location.y() }, scaled_bitmap, scaled_bitmap->rect()); */
```
[the remark this answers is not in the document]
Comment (tslater2006) [reply Userland/Applications/PixelPaint/Tools/TextTool.cpp]: Old code I forgot to remove before opening the PR :(. Cleaned up!

File: Userland/Applications/PixelPaint/Tools/TextTool.cpp (hunk 4 above)
Comment (tslater2006) [reply Userland/Applications/PixelPaint/Tools/TextTool.cpp]: This was actually pretty easy! Thanks for having me take another look at it. The newest code lets TextEditor say where the cursor is, we then scale it based on ImageEditor scale() and then position it where the text is on screen and draw. About as many lines due needing to scale an IntRect by a float, but certainly easier to follow than before.

**Comment by tslater2006** [comment]
Latest push addresses almost all feedback, and fixed the drawing on top of the rulers. Outstanding feedback has comments on it :)

File: Userland/Applications/PixelPaint/Tools/TextTool.cpp [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+
+    // Draw the blinking cursor.
+    if (m_cursor_blink_state) {
+        auto editor_cursor_rect = m_text_editor->cursor_content_rect();
+
+        /* TextEditor starts left most at 3, for TextTool this ends up putting the cursor in the middle of the letter.
+         * Looks better if we treat 0 as left most here, so we just translate it to the left. */
+        editor_cursor_rect.translate_by(-3, 0);
+
+        /* ImageEditor scale is a float, but we are working with int and IntRects. */
```
Comment (AtkinsSJ) [inline Userland/Applications/PixelPaint/Tools/TextTool.cpp:220]: Nit-pick: Use `// ...` for comments, instead of `/* ... */`

**Review by AtkinsSJ — CHANGES_REQUESTED** [review]
One last(?) tiny nit-pick.

File: Userland/Applications/PixelPaint/Tools/TextTool.cpp (hunk 11 above)
Comment (tslater2006) [reply Userland/Applications/PixelPaint/Tools/TextTool.cpp]: Should be fixed now with latest push.

File: Userland/Applications/PixelPaint/ImageEditor.cpp [hunk 12]
```diff
@@ -264,8 +264,17 @@ Gfx::IntRect ImageEditor::mouse_indicator_rect_y() const
...
+    if (m_active_tool) {
+        if (m_show_rulers) {
+            auto clipped_event = GUI::PaintEvent(Gfx::IntRect { event.rect().x() + m_ruler_thickness,
+                                                     event.rect().y() + m_ruler_thickness,
+                                                     event.rect().width() - m_ruler_thickness,
+                                                     event.rect().height() - m_ruler_thickness },
+                event.window_size());
+            m_active_tool->on_second_paint(m_active_layer, clipped_event);
+        } else
+            m_active_tool->on_second_paint(m_active_layer, event);
```
Comment (alimpfard) [inline Userland/Applications/PixelPaint/ImageEditor.cpp:276]: Please put braces around both branches.

File: Userland/Applications/PixelPaint/Tools/TextTool.cpp [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+            if (picker->exec() == GUI::Dialog::ExecResult::OK) {
+                m_font_label->set_text(picker->font()->human_readable_name());
+                m_selected_font = picker->font();
+                m_text_editor->set_font(m_selected_font);
+                m_editor->set_focus(true);
+            }
+        };
+    }
+
+    return m_properties_widget.ptr();
```
Comment (alimpfard) [inline Userland/Applications/PixelPaint/Tools/TextTool.cpp:120]: Stylistic suggestion (feel free to ignore) - the diff looks odd, so move the if negated to the top, dedent the rest.

```suggestion
    if (m_properties_widget)
        return m_properties_widget.ptr();
    
    m_properties_widget = GUI::Widget::construct();
    m_properties_widget->set_layout<GUI::VerticalBoxLayout>();

    auto& font_header = m_properties_widget->add<GUI::Label>("Current Font:");
    font_header.set_text_alignment(Gfx::TextAlignment::CenterLeft);

    m_font_label = m_properties_widget->add<GUI::Label>(m_selected_font->human_readable_name());

    auto& change_font_button = m_properties_widget->add<GUI::Button>("Change Font...");
    change_font_button.on_click = [&](auto) {
        auto picker = GUI::FontPicker::construct(nullptr, m_selected_font, false);
        if (picker->exec() == GUI::Dialog::ExecResult::OK) {
            m_font_label->set_text(picker->font()->human_readable_name());
            m_selected_font = picker->font();
            m_text_editor->set_font(m_selected_font);
            m_editor->set_focus(true);
        }
    };
    return m_properties_widget.ptr();
```

File: Userland/Applications/PixelPaint/Tools/TextTool.cpp [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+    m_text_editor->handle_keyevent({}, event);
+    m_editor->update();
+    return true;
+}
+
+Variant<Gfx::StandardCursor, NonnullRefPtr<Gfx::Bitmap>> TextTool::cursor()
+{
+    if (m_mouse_is_over_text)
+        return Gfx::StandardCursor::Move;
+    else
```
Comment (alimpfard) [inline Userland/Applications/PixelPaint/Tools/TextTool.cpp:291]: Pls drop this else.

**Review by alimpfard — COMMENTED** [review]
Just some smol comments.

File: Userland/Applications/PixelPaint/Tools/TextTool.cpp (hunk 14 above)
Comment (tslater2006) [reply Userland/Applications/PixelPaint/Tools/TextTool.cpp:291]: Not sure I understand why to drop the else. In the event you are not over the text we want the cursor to be the arrow.

File: Userland/Applications/PixelPaint/Tools/TextTool.cpp (hunk 13 above)
Comment (tslater2006) [reply Userland/Applications/PixelPaint/Tools/TextTool.cpp]: This is following the pattern that is in the other tools but I agree it looks better the other way around. I've made this change in the latest push

File: Userland/Applications/PixelPaint/Tools/TextTool.cpp (hunk 14 above)
Comment (tslater2006) [reply Userland/Applications/PixelPaint/Tools/TextTool.cpp:291]: Doh! I get it now. No need for the else, just have it be 

```
    if (m_mouse_is_over_text)
        return Gfx::StandardCursor::Move;
    
    return Gfx::StandardCursor::Arrow;
    ```

**Review by AtkinsSJ — APPROVED** [review]
Let's get this party started!

**Pull request merged**